### PR TITLE
Chapter3/homework/10

### PR DESCRIPTION
**Pull request by matthew-hole**
Solution to Homework 10. 

Initially overcomplicated the creating of the TextTable but after that it was fairly straightforward. Used your example from the live coding part of the lesson 'AttributeGetter' as I was initially unsure how I was to retrieve the 'attributes'.

Nevertheless, homework is done!

### CODE REVIEW

File: Src/BootCamp.Chapter/TextBoxPrinter.cs [hunk 1]
```diff
@@ -1,12 +1,58 @@
...
+                    sb.AppendLine(paddingLine);
+                }
+                else if (i == numberOfMainLine && textTableAttributes.Padding == 0)
+                {
+                    sb.AppendLine(mainLine);
+                }
+                else if (i == numberOfMainLine && textTableAttributes.Padding != 0)
+                {
+                    sb.AppendLine(mainLineWithPadding);
+                }
```
Comment (matthew-hole) [inline Src/BootCamp.Chapter/TextBoxPrinter.cs]: All of this above could probably be improved through refactoring into its own private method to reduce the complexity in the Print() method. Let me know if you spot any other ways of achieving this.

File: Src/BootCamp.Chapter/TextBoxPrinter.cs [hunk 2]
```diff
@@ -1,12 +1,58 @@
...
+
+            var table = GenerateTable(getTextTableAttribute, obj.ToString());
+            Console.WriteLine(table);
+
+            return GenerateTable(getTextTableAttribute, obj.ToString());
+        }
+
+        private static string GenerateTable(TextTableAttribute textTableAttributes, string message)
+        {
+            int messageLength = message.Length + (textTableAttributes.Padding * 2);
```
Comment (Almantask) [inline Src/BootCamp.Chapter/TextBoxPrinter.cs]: Get rid of magic numbers 😄

File: Src/BootCamp.Chapter/TextBoxPrinter.cs (hunk 1 above)
Comment (Almantask) [reply Src/BootCamp.Chapter/TextBoxPrinter.cs]: Maybe not only private methods, maybe a class of its' own. Builder pattern and fluent interface sounds like a good fit.

Also, a lot of redundancies are spotted.

**Review by Almantask — CHANGES_REQUESTED** [review]

File: Src/BootCamp.Chapter/TextBoxPrinter.cs (hunk 2 above)
Comment (matthew-hole) [reply Src/BootCamp.Chapter/TextBoxPrinter.cs]: Unsure how to remove the magic numbers when they are fundamental to the logic used? What is the best practice for achieving this?

Comment (Almantask) [reply Src/BootCamp.Chapter/TextBoxPrinter.cs]: Instead of magic numbers, used named constants.

**Comment by matthew-hole** [comment]
@Almantask Please see commit "f5fc7c1" for changes relating to your comments. 

Note that there is still some complexity with multiple else if statements. I spent a while researching state-machines/state-pattern and now understand its uses and deemed it was not sensible to be used here. 

PR should be good to go! Thanks.

File: Src/BootCamp.Chapter/TextBoxBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    public class TextBoxBuilder
+    {
+        private readonly string _message;
+        private readonly TextTableAttribute _textTableAttributes;
+        private readonly int _messageLength;
+        private readonly int _numberOfLines;
+        private readonly int _numberOfMainLine;
+        private readonly string _paddingString;
+        private StringBuilder sb = new StringBuilder();
+        private const int DivideByTwo = 2;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/TextBoxBuilder.cs:15]: DivideByTwo or Addone does not make anything better. You need to describe those constants in a logical why (what are they used for?). 

You did good with `NumberOfLinesWithoutPadding`

**Review by Almantask — APPROVED** [review]

**Review by Almantask — APPROVED** [review]
Yeah, I don't see much to pick on. Carry on 👍

**Pull request closed without merging**